### PR TITLE
Refactor netebpfext in preparation for fuzz testing

### PR DESCRIPTION
**Pull request by dthaler**
## Description

This PR allows netebpfext to have a static lib that can be tested in user mode as well as compiled into the kernel driver

This PR does contain one ugly workaround but I haven't found a better method.  Specifically, the netebpfext_user and netebpfext_unit projects both have "$(WindowsSdkDir)Include\10.0.22000.0\km" in the include path in order to find the DDK includes from an SDK project.  Github runners seem to have 10.0.22621.0 of the SDK installed but 10.0.22000.0 of the DDK so simply doing "..\km" was not sufficient for github to build so this workaround ties the file to a specific DDK version and the vcxproj files will need to be updated once github upgrades to a later DDK but in the meantime this seems to work fine to find the headers.

## Testing

This is needed to unblock work for issue #1325 

## Documentation

No impact

### CODE REVIEW

**Review by Alan-Jowett — APPROVED** [review]

File: netebpfext/user/fwp_um.cpp [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+        return FWP_ACTION_CALLOUT_UNKNOWN;
+    }
+    FWPS_FILTER fwps_filter = {.context = fwpm_filter->rawContext};
+    NET_BUFFER_LIST_POOL_PARAMETERS pool_parameters = {};
+    HANDLE nbl_pool_handle = NdisAllocateNetBufferListPool(nullptr, &pool_parameters);
+    if (!nbl_pool_handle) {
+        return FWP_ACTION_CALLOUT_UNKNOWN;
+    }
+    NET_BUFFER_LIST* nbl = NdisAllocateNetBufferList(nbl_pool_handle, 0, 0);
+    if (nbl) {
```
Comment (shankarseal) [inline netebpfext/user/fwp_um.cpp:38]: Dont care about alloc failures?

File: netebpfext/user/fwp_um.cpp [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+}
+
+_IRQL_requires_max_(PASSIVE_LEVEL) NTSTATUS
+    FwpsCalloutRegister3(_Inout_ void* device_object, _In_ const FWPS_CALLOUT3* callout, _Out_opt_ uint32_t* callout_id)
+{
+    UNREFERENCED_PARAMETER(device_object);
+
+    auto& engine = *_fwp_engine::get()->get();
+
+    auto id_returned = engine.add_fwps_callout(callout);
```
Comment (shankarseal) [inline netebpfext/user/fwp_um.cpp:128]: nit: rename to `register_callout`.

File: netebpfext/user/fwp_um.h [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        for (auto& [first, filter] : fwpm_filters) {
+            if (filter.rawContext != 0) {
+                return &filter;
+            }
+        }
+        return nullptr;
+    }
+
+    _Ret_maybenull_ const GUID*
+    get_callout_key_from_layer_guid(_In_ const GUID* layer_guid)
```
Comment (shankarseal) [inline netebpfext/user/fwp_um.h:101]: Why is the assumption made that there will be one callout per layer?

File: netebpfext/user/fwp_um.h [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    }
+
+    static std::unique_ptr<_fwp_engine> _engine;
+
+    std::mutex lock;
+    uint32_t next_id = 1;
+    std::unordered_map<size_t, FWPS_CALLOUT3> fwps_callouts;
+    std::unordered_map<size_t, FWPM_CALLOUT0> fwpm_callouts;
+    std::unordered_map<size_t, FWPM_FILTER0> fwpm_filters;
+    std::unordered_map<size_t, FWPM_SUBLAYER0> fwpm_sub_layers;
```
Comment (shankarseal) [inline netebpfext/user/fwp_um.h:129]: I dont think we need this. The sublayer functions can simply return incremented id without actually storing any object.

File: netebpfext/user/kernel_um.cpp [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+        (mdl)->next = (PMDL)NULL;                                                                                 \
+        (mdl)->size =                                                                                             \
+            (uint16_t)(sizeof(MDL) + (sizeof(PFN_NUMBER) * ADDRESS_AND_SIZE_TO_SPAN_PAGES((base_va), (length)))); \
+        (mdl)->flags = 0;                                                                                         \
+        (mdl)->start_va = (PVOID)PAGE_ALIGN((base_va));                                                           \
+        (mdl)->byte_offset = BYTE_OFFSET((base_va));                                                              \
+        (mdl)->byte_count = (ULONG)(length);                                                                      \
+    }
+
+unsigned long __cdecl DbgPrintEx(
```
Comment (shankarseal) [inline netebpfext/user/kernel_um.cpp:68]: This is a relic from the early prototype days. We may want to replace this with Tracelog.

File: netebpfext/user/kernel_um.cpp [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+    KeAcquireSpinLockRaiseToDpc(_Inout_ PKSPIN_LOCK spin_lock)
+{
+    auto lock = reinterpret_cast<ebpf_lock_t*>(spin_lock);
+    return ebpf_lock_lock(lock);
+}
+
+_Requires_lock_held_(*spin_lock) _Releases_lock_(*spin_lock) _IRQL_requires_(DISPATCH_LEVEL) void KeReleaseSpinLock(
+    _Inout_ PKSPIN_LOCK spin_lock, _In_ _IRQL_restores_ KIRQL new_irql)
+{
+    auto lock = reinterpret_cast<ebpf_lock_t*>(spin_lock);
```
Comment (shankarseal) [inline netebpfext/user/kernel_um.cpp:280]: we can do this later, but netebpfext need not use any ebpf core platform lib.

File: netebpfext/user/kernel_um.h [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+
+NTSTATUS
+RtlULongAdd(
+    _In_ unsigned long augend,
+    _In_ unsigned long addend,
+    _Out_ _Deref_out_range_(==, augend + addend) unsigned long* result);
+
+// Inline functions
+_Must_inspect_result_ BOOLEAN CFORCEINLINE
+IsListEmpty(_In_ const LIST_ENTRY* list_head)
```
Comment (shankarseal) [inline netebpfext/user/kernel_um.h:212]: If `LIST_ENTRY` struct was obtained from some header, why couldnt the inline functions too be obtained from there?

File: netebpfext/user/nmr_impl.cpp [hunk 8]
```diff
@@ -0,0 +1,353 @@
...
+        return std::nullopt;
+    }
+
+    // Skip if client or provider are deregestering.
+    if (client.deregistering || provider.deregistering) {
+        return std::nullopt;
+    }
+
+    // Acquire references on both client and provider to prevent them from unloading.
+    client.bindings++;
```
Comment (shankarseal) [inline netebpfext/user/nmr_impl.cpp:166]: No interlocked op?

File: tests/netebpfext_unit/netebpf_ext_helper.cpp [hunk 9]
```diff
@@ -0,0 +1,115 @@
+// Copyright (c) Microsoft Corporation
+// SPDX-License-Identifier: MIT
+
+#include "netebpf_ext_helper.h"
+
+ebpf_registry_key_t ebpf_root_registry_key = HKEY_CURRENT_USER;
+DEVICE_OBJECT* _net_ebpf_ext_driver_device_object;
+
+_netebpf_ext_helper::_netebpf_ext_helper()
```
Comment (shankarseal) [inline tests/netebpfext_unit/netebpf_ext_helper.cpp:9]: Can this be refactored into a common code lib?

File: netebpfext/user/fwp_um.h [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+    get()
+    {
+        if (!_engine)
+            _engine = std::make_unique<_fwp_engine>();
+        return _engine.get();
+    }
+
+  private:
+    _Ret_maybenull_ const FWPM_FILTER*
+    get_fwpm_filter_with_context()
```
Comment (shankarseal) [inline netebpfext/user/fwp_um.h:90]: Why return arbitrary filter here? Probably depending on what we are testing, a specific filter should be returned as the hook providers make use of the filter context, isnt it?

File: netebpfext/user/kernel_um.cpp (hunk 5 above)
Comment (Alan-Jowett) [reply netebpfext/user/kernel_um.cpp:68]: It's used by KdPrintEx. I propose we keep the switch from KdPrintEx -> tracelog as a separate PR.

Comment (Alan-Jowett) [reply netebpfext/user/kernel_um.cpp:68]: https://github.com/microsoft/ebpf-for-windows/issues/1466

File: netebpfext/user/kernel_um.h (hunk 7 above)
Comment (Alan-Jowett) [reply netebpfext/user/kernel_um.h:212]: LIST_ENTRY and IsListEmpty are in different header files.
LIST_ENTRY is in winnt.h
IsListEmpty is in wdm.h

wdm.h conflicts with many of the user mode headers and can't be included, hence the duplication of the functions we need.

File: tests/netebpfext_unit/netebpf_ext_helper.cpp (hunk 9 above)
Comment (Alan-Jowett) [reply tests/netebpfext_unit/netebpf_ext_helper.cpp:9]: Which part of it?

File: netebpfext/user/fwp_um.h (hunk 10 above)
Comment (Alan-Jowett) [reply netebpfext/user/fwp_um.h:90]: Agreed. What is it supposed to match on here? Is it layer_guid against FWPM_FILTER0::layerKey?

File: tests/netebpfext_unit/netebpf_ext_helper.cpp (hunk 9 above)
Comment (Alan-Jowett) [reply tests/netebpfext_unit/netebpf_ext_helper.cpp:9]: Propose we refactor it into common code should the need arise.

File: netebpfext/user/fwp_um.h (hunk 10 above)
Comment (Alan-Jowett) [reply netebpfext/user/fwp_um.h:90]: https://github.com/microsoft/ebpf-for-windows/issues/1474

**Review by Alan-Jowett — APPROVED** [review]

**Comment by dthaler** [comment]
I've reviewed Alan's subsequent commits to this PR

**Review by shankarseal — APPROVED** [review]

**Pull request merged**